nav2_behavior_tree/include/nav2_behavior_tree/bt_action_server.hpp

```diff
@@ -0,0 +1,392 @@
+// Copyright (c) 2020 Sarthak Mittal
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef NAV2_BEHAVIOR_TREE__BT_ACTION_SERVER_HPP_
+#define NAV2_BEHAVIOR_TREE__BT_ACTION_SERVER_HPP_
+
+#include <memory>
+#include <string>
+#include <vector>
+#include <fstream>
+#include <set>
+#include <exception>
+
+#include "geometry_msgs/msg/pose_stamped.hpp"
+#include "nav2_behavior_tree/behavior_tree_engine.hpp"
+#include "nav2_behavior_tree/ros_topic_logger.hpp"
+#include "nav2_util/lifecycle_node.hpp"
+#include "nav2_util/simple_action_server.hpp"
+#include "rclcpp_action/rclcpp_action.hpp"
+#include "tf2_ros/buffer.h"
+#include "tf2_ros/transform_listener.h"
+#include "tf2_ros/create_timer_ros.h"
+
+namespace nav2_behavior_tree
+{
+/**
+ * @class nav2_behavior_tree::BtActionServer
+ * @brief An action server that uses behavior tree to execute an action
+ */
+template<class ActionT>
+class BtActionServer
```

Comment:
I'm having some trouble linking the nav2_behavior_tree library while building nav2_bt_navigator when I split hpp / cpp.

I'm adding `src/bt_action_server.cpp` here:
https://github.com/ros-planning/navigation2/blob/6a0c92cad7fb8bb1e3e8500742bc293cddbedc36/nav2_behavior_tree/CMakeLists.txt#L46-L48

Since the library installs and exports are already taken care of, I assumed `nav2_bt_navigator` would build without any issues but I end up getting this error:
`libbt_navigator_core.so: undefined reference to nav2_behavior_tree::BtActionServer<nav2_msgs::action::NavigateToPose>`

Any idea what I'm doing incorrectly here?

Reply:
You need to create supported types in the library by creating instances of the valid template types https://github.com/ros-planning/navigation2/blob/main/nav2_smac_planner/src/a_star.cpp#L647.

... but that does bring up the issue of what happens when other users want to add them. Maybe just a header is OK,but I'd like the class definition separate from the implementations within the header. Its useful to have the class compactly so you can review the API without seeing all the function implementations.

It's also possible that the instantiation can happen elsewhere, like in the header before an application class, I'm not 100% sure off  the top of my head. That would make it possible to use other messages types and split the files.

Reply:
Going through some rclcpp code, looks like the way to do this would be to shift the implementation to a new header file, call it `bt_action_server_impl.hpp` and include it in `bt_action_server.hpp`. (refer https://github.com/ros2/rclcpp/blob/master/rclcpp_action/include/rclcpp_action/client_goal_handle_impl.hpp)

Reply:
that also works for me. What's the functional difference though between calling it a `hpp` or a `cpp` in another directory though?